pkg/clustermanager/discovery.go

```diff
@@ -61,28 +62,55 @@ func (c *controller) CreateProvider(
 	if err != nil {
 		return nil, err
 	}
-	c.providers.providerList[providerName] = newProvider
+	c.providers[providerName] = providerInfo{providerClient: newProvider}
 	return newProvider, nil
 }
 
-func (c *controller) StartDiscovery(
-	provider clusterproviderclient.ProviderClient,
-	providerName string) error {
-	w, err := provider.StartWatch()
+// StartDiscovery will start watching provider clusters for changes
+func (c *controller) StartDiscovery(providerName string) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	providerInfo, ok := c.providers[providerName]
+	if !ok {
+		return errors.New("failed to start provider discovery. provider info not exists")
+	}
+
+	watcher, err := providerInfo.providerClient.Watch()
 	if err != nil {
 		return err
 	}
-	go ProcessProviderWatchEvents(c.context, w, c.clientset, providerName)
+	go processProviderWatchEvents(c.context, watcher, c.clientset, providerName)
+
+	providerInfo.providerWatcher = watcher
+	c.providers[providerName] = providerInfo
+	return nil
+}
+
+// StopDiscovery will stop watching provider clusters for changes
+func (c *controller) StopDiscovery(providerName string) error {
```

Comment:
Where do we use it? Or is it going to be on next PRs ?

Reply:
next PR. we will use it on provider reconcile.